ddapm_test_agent/agent.py

```diff
@@ -345,18 +349,34 @@ def _requests_by_session(self, token: Optional[str]) -> List[Request]:
         """
         # Go backwards in the requests received gathering requests until
         # the /session-start request for the token is found.
-        reqs: List[Request] = []
+        # Note that this may not return all associated traces, because some
+        # may be generated before the session-start call
+        session_reqs: List[Request] = []
+        sessionless_reqs: List[Request] = []
+        matched = token is None
+
         for req in reversed(self._requests):
             if req.match_info.handler == self.handle_session_start:
-                if token is None or _session_token(req) == token:
+                if token is None:
+                    # If no token is specified, then we match the latest session
                     break
-                else:
-                    # The requests made were from a different manual session
-                    # so continue.
+                elif _session_token(req) == token:
+                    # If a token is specified and it matches, we've hit the start of our session
+                    matched = True
+                    break
+                elif _session_token(req) != token:
+                    # If a token is specified and it doesn't match, we've hit the start of a different session
+                    # So we reset the list of requests
+                    sessionless_reqs: List[Request] = []
                     continue
-            if _session_token(req) in [token, None]:
-                reqs.append(req)
-        return reqs
+            if _session_token(req) == token:
+                session_reqs.append(req)
+            elif _session_token(req) is None:
+                sessionless_reqs.append(req)
+
+        if not matched and not session_reqs:
```

Comment:
Here's the "session created or we have at last one request with the session" logic